### DISCUSSION TITLE
CAN-Bus on OrangePi 5 Plus

### DISCUSSION THREAD

**Discussion by L1800Turbo** [Joshua-Riek/ubuntu-rockchip#918] [Q&A]
Hello,
I had an Orange Pi 5 Plus running with the Orangepi OS and switched to this Ubuntu-Rockchip version now.
On the OrangePi OS I was able to activate the two CAN interfaces through the orangepi-config tool. 

Did someone already activate it here? 
Unfortunately I don't remember the drivers, I'd have to look that up. But I assume, there is some configuration on the pinmux necessary to configure the IOs for CAN.

Thanks

**Comment by Joshua-Riek** [comment]
This is not an Armbian fork, so there is no orangepi-config tool. Read the GitHub wiki page and look at the device tree overlays.

**Comment by L1800Turbo** [comment]
Ah that's a good hint, thank you!
I didn't expect a orangepi-config tool, that's just the way I did once before.

Do I also need a rk3588-can driver or something similar? I didn't find such driver preinstalled so far.

**Reply by Joshua-Riek** [reply]
Drivers are installed. You need a device tree overlay.

**Comment by L1800Turbo** [comment]
Okay, then I might have misconfigured it.
I uncommented in `/etc/default/u-boot` the FDT values and filled like below:
```
U_BOOT_FDT="device-tree/rockchip/rk3588-orangepi-5-plus.dtb"
U_BOOT_FDT_OVERLAYS="device-tree/rockchip/overlay/rk3588-can0-m0.dtbo"
```
followed by u-boot-update and reboot.

`/boot/extlinux/extlinux.conf` got the fdt params filled in with proper paths..
Did I forget anything?

**Reply by Joshua-Riek** [reply]
You can comment out `U_BOOT_FDT`, but this looks correct. If its not working post the contents of /boot/extlinux/extlinux.conf

**Reply by L1800Turbo** [reply]
I had `U_BOOT_FDT` after it didn't work, as I wasn't sure if it's needed.

This is the extlinux.conf:
```
## /boot/extlinux/extlinux.conf
##
## IMPORTANT WARNING
##
## The configuration of this file is generated automatically.
## Do not edit this file manually, use: u-boot-update

default l0
menu title U-Boot menu
prompt 1
timeout 20


label l0
        menu label Ubuntu 24.04 LTS 6.1.0-1019-rockchip
        linux /boot/vmlinuz-6.1.0-1019-rockchip
        initrd /boot/initrd.img-6.1.0-1019-rockchip
        fdtdir /lib/firmware/6.1.0-1019-rockchip/device-tree/
        fdtoverlays  /lib/firmware/6.1.0-1019-rockchip/device-tree/rockchip/overlay/rk3588-can0-m0.dtbo
        append root=UUID=5ea6bcd5-8d10-4274-8cd6-a843f6587e89 rootwait rw console=ttyS2,1500000 console=tty1 cgroup_enable=cpuset cgroup_memory=1 cgroup_enable=memory

label l0r
        menu label Ubuntu 24.04 LTS 6.1.0-1019-rockchip (rescue target)
        linux /boot/vmlinuz-6.1.0-1019-rockchip
        initrd /boot/initrd.img-6.1.0-1019-rockchip
        fdtdir /lib/firmware/6.1.0-1019-rockchip/device-tree/
        append root=UUID=5ea6bcd5-8d10-4274-8cd6-a843f6587e89 rootwait rw console=ttyS2,1500000 console=tty1 cgroup_enable=cpuset cgroup_memory=1 cgroup_enable=memory single


label l1
        menu label Ubuntu 24.04 LTS 6.1.0-1016-rockchip
        linux /boot/vmlinuz-6.1.0-1016-rockchip
        initrd /boot/initrd.img-6.1.0-1016-rockchip
        fdtdir /lib/firmware/6.1.0-1016-rockchip/device-tree/
        fdtoverlays  /lib/firmware/6.1.0-1016-rockchip/device-tree/rockchip/overlay/rk3588-can0-m0.dtbo
        append root=UUID=5ea6bcd5-8d10-4274-8cd6-a843f6587e89 rootwait rw console=ttyS2,1500000 console=tty1 cgroup_enable=cpuset cgroup_memory=1 cgroup_enable=memory

label l1r
        menu label Ubuntu 24.04 LTS 6.1.0-1016-rockchip (rescue target)
        linux /boot/vmlinuz-6.1.0-1016-rockchip
        initrd /boot/initrd.img-6.1.0-1016-rockchip
        fdtdir /lib/firmware/6.1.0-1016-rockchip/device-tree/
        append root=UUID=5ea6bcd5-8d10-4274-8cd6-a843f6587e89 rootwait rw console=ttyS2,1500000 console=tty1 cgroup_enable=cpuset cgroup_memory=1 cgroup_enable=memory single

```

**Reply by L1800Turbo** [reply]
I did some additional debugging, but it seems to me, that the module is loaded correctly. However, my expectation was that the kernel would detect the CAN module then.

I connected the serial debug, because I wasn't sure about the paths in the bootloader, but the overlay seems to be loaded:
```
Device 0: Vendor: 0x15b7 Rev: B35900WD Prod: 172844424547
            Type: Hard Disk
            Capacity: 488386.3 MB = 476.9 GB (1000215216 x 512)
... is now current device
Scanning nvme 0:2...
Found /boot/extlinux/extlinux.conf
Retrieving file: /boot/extlinux/extlinux.conf
1917 bytes read in 4 ms (467.8 KiB/s)
U-Boot menu
1:      Ubuntu 24.04 LTS 6.1.0-1019-rockchip
2:      Ubuntu 24.04 LTS 6.1.0-1019-rockchip (rescue target)
3:      Ubuntu 24.04 LTS 6.1.0-1016-rockchip
4:      Ubuntu 24.04 LTS 6.1.0-1016-rockchip (rescue target)
Enter choice: 1:        Ubuntu 24.04 LTS 6.1.0-1019-rockchip
Retrieving file: /boot/initrd.img-6.1.0-1019-rockchip
16271227 bytes read in 45 ms (344.8 MiB/s)
Retrieving file: /boot/vmlinuz-6.1.0-1019-rockchip
40962560 bytes read in 101 ms (386.8 MiB/s)
append: root=UUID=5ea6bcd5-8d10-4274-8cd6-a843f6587e89 rootwait rw console=ttyS2,1500000 console=tty1 cgroup_enable=cpuset cgroup_memory=1 cgroup_enable=memory
Retrieving file: /lib/firmware/6.1.0-1019-rockchip/device-tree/rockchip/rk3588-orangepi-5-plus.dtb
281782 bytes read in 86 ms (3.1 MiB/s)
Retrieving file: /lib/firmware/6.1.0-1019-rockchip/device-tree/rockchip/overlay/rk3588-can0-m0.dtbo
343 bytes read in 119 ms (2 KiB/s)
Fdt Ramdisk skip relocation
## Flattened Device Tree blob at 0x0a100000
   Booting using the fdt blob at 0x0a100000
   reserving fdt memory region: addr=a100000 size=47000
  'reserved-memory' cma: addr=10000000 size=10000000
  'reserved-memory' ramoops@110000: addr=110000 size=e0000
   Using Device Tree in place at 000000000a100000, end 000000000a149fff
Adding bank: 0x00200000 - 0xf0000000 (size: 0xefe00000)
Adding bank: 0x100000000 - 0x3fc000000 (size: 0x2fc000000)
Adding bank: 0x3fc500000 - 0x3fff00000 (size: 0x03a00000)
Adding bank: 0x4f0000000 - 0x500000000 (size: 0x10000000)
Total: 6274.213 ms
```

Any additional suggestions?

**Reply by L1800Turbo** [reply]
I took out the NVMe today and put in a flash card with the OrangePi 1.0.8 Jammy to compare the boot logs:
```
mmc0 is current device
Scanning mmc 0:1...
Found U-Boot script /boot.scr
reading /boot.scr
3411 bytes read in 4 ms (832 KiB/s)
## Executing script at 00500000
Boot script loaded from mmc 0
reading /orangepiEnv.txt
244 bytes read in 3 ms (79.1 KiB/s)
reading /uInitrd
19967902 bytes read in 1607 ms (11.8 MiB/s)
reading /Image
34327040 bytes read in 2758 ms (11.9 MiB/s)
reading /dtb/rockchip/rk3588-orangepi-5-plus.dtb
266904 bytes read in 28 ms (9.1 MiB/s)
reading /dtb/rockchip/overlay/rk3588-can0-m0.dtbo
343 bytes read in 6 ms (55.7 KiB/s)
Applying kernel provided DT overlay rk3588-can0-m0.dtbo
reading /dtb/rockchip/overlay/rk3588-fixup.scr
2756 bytes read in 6 ms (448.2 KiB/s)
Applying kernel provided DT fixup script (rk3588-fixup.scr)
## Executing script at 09000000
Fdt Ramdisk skip relocation
```
I see an additional fixup (don't know if that's relevant) and the line `Applying kernel provided DT overlay rk3588-can0-m0.dtbo`. This seems to come out of the `boot.cmd` startup script:
```
load ${devtype} ${devnum} ${ramdisk_addr_r} ${prefix}uInitrd
load ${devtype} ${devnum} ${kernel_addr_r} ${prefix}Image

load ${devtype} ${devnum} ${fdt_addr_r} ${prefix}dtb/${fdtfile}
fdt addr ${fdt_addr_r}
fdt resize 65536
for overlay_file in ${overlays}; do
	if load ${devtype} ${devnum} ${load_addr} ${prefix}dtb/rockchip/overlay/${overlay_prefix}-${overlay_file}.dtbo; then
		echo "Applying kernel provided DT overlay ${overlay_prefix}-${overlay_file}.dtbo"
		fdt apply ${load_addr} || setenv overlay_error "true"
	fi
done
for overlay_file in ${user_overlays}; do
	if load ${devtype} ${devnum} ${load_addr} ${prefix}overlay-user/${overlay_file}.dtbo; then
		echo "Applying user provided DT overlay ${overlay_file}.dtbo"
		fdt apply ${load_addr} || setenv overlay_error "true"
	fi
done
```
A question for me would be, if some kind of apply also happens with extlinux.conf and I just don't see it in the logs.

I also noticed PCIe problems. But this doesn't seem to be the root I guess...
```
[   10.785600] rk-pcie fe170000.pcie: PCIe Link Fail, LTSSM is 0x3, hw_retries=0
[   14.805621] rk-pcie fe170000.pcie: PCIe Link Fail, LTSSM is 0x3, hw_retries=1
[   15.892301] rk-pcie fe170000.pcie: failed to initialize host
```

Is this U-Boot version the right one: `U-Boot 2017.09 (Jun 12 2024 - 21:14:05 +0000)`
I flashed it from the running system with `sudo u-boot-install-mtd`

Any ideas would help!

**Comment by randellhodges** [comment]
@L1800Turbo Did you get this working? I have an Orange PI 5 Max in the mail and I'm hoping to use canbus with it. I noticed in Orange Pi's documentation, they state that can works with their 5 series kernel, but not their 6 series. I'm new to all this, so I haven't a clue if that means there are other factors in play beyond just overlays.

Hopefully in week or 2 I'll have the board in hand to play with. I'm expecting to learn a lot.

**Reply by L1800Turbo** [reply]
Yes, at least for a first test I got it working.
Other than stated it seems that there actually _are_ drivers missing: `rockchip_can` and `rockchip_canfd`.

I compiled them using the kernel sources and manually added the drivers using `modprobe`.
A first `candump` try worked.

I planned to look that the module is also available after a kernel update (maybe by including it by default or automatically compiling).

**Reply by Healthyyue** [reply]
I found the rockchip_can and rockchip_canfd driver in the system(release 2.4.0),but after I using modprobe and modify the /etc/default/u-boo as below.The can bus is still not work.
 may I ask  which version of kernel source that you compile these two drivers?

**Reply by Healthyyue** [reply]
I’m not sure do you need the solution still. I have make the can work well on orange Pi 5 max.
We don't need to compile the driver,just only set the device tree overlays right well make it work.
1. make sure you have download the release 2.4.0 of ubuntu-rockchip.
2. modify the /etc/default/u-boot:
```
	U_BOOT_FDT="/usr/lib/firmware/5.10.0-1012-rockchip/device-tree/rockchip/rk3588-orangepi-5-max.dtb"
	U_BOOT_FDT_OVERLAYS="rk3588-can0-m0.dtbo rk3588-can1-m0.dtbo"
	U_BOOT_FDT_OVERLAYS_DIR="/usr/lib/firmware/5.10.0-1012-rockchip/device-tree/rockchip/overlay"
```

3. run commands like this:
```
sudo u-boot-update


sudo reboot
```

4. check the CAN
```
sudo ifconfig -a
```

**Reply by L1800Turbo** [reply]
In my setup I used the 2.2.1 (and still running today).

After getting the kernel sources (https://github.com/Joshua-Riek/linux-rockchip.git) and compiling the can driver all I needed to change/uncomment in `/etc/default/u-boot` was one line:
`U_BOOT_FDT_OVERLAYS="device-tree/rockchip/overlay/rk3588-can0-m0.dtbo device-tree/rockchip/overlay/rk3588-can1-m0.dtbo device-tree/rockchip/overlay/rk3588-i2c4-m3.dtbo device-tree/rockchip/overlay/rk3588-uart6-m1.dtbo"`

With `ifconfig` I was able to activate CAN after a restart.

So in the 2.4.0 you integrated the CAN drivers?
That would be good to know for an update. Then I wouldn't need to think about automatic modules myself.

**Reply by Healthyyue** [reply]
Yes, the drivers about can are installed in 2.4.0.

**Reply by jagauthier** [reply]
Greetings!  I just got an OrangePi 5 Plus and I want to use it with CAN.  I found this thread, and after applying the overlay I do not see the devices.  I'm wondering if I need to install the drivers that were mentioned.   Is that confirmed?
